File: Data_Structure/C/DoublyCircularLL.c
// Doubly Circular Linked List
#include<stdio.h>
#include<stdlib.h>

#pragma pack(1)
struct node 
{
    int data;
    struct node *next;
    struct node *previous;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void Display(PNODE first, PNODE last)
{
    // Input filter
    if(first == NULL && last == NULL)
    {
        return;
    }

    printf(" <=> ");

    do 
    {
        printf("| %d | <=> ",first->data);
        first = first->next;

    }while(first != last->next);

    printf("\n");

}

int Count(PNODE first, PNODE last)
{
    int iCount = 0;

    if(first == NULL && last == NULL)
    {
        return iCount;
    }

    do 
    {
        iCount++;
        first = first->next;

    }while(first != last->next);


    return iCount;
}

void InsertFirst(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = iNo;
    newn->next = NULL;
    newn->previous = NULL;

    if(*first == NULL && *last == NULL)     
    {
        *first = newn;
        *last = newn;
    }

    else                              
    {
        newn->next = *first;
        (*first)->previous = newn;
        *first = newn;
    }

    (*last)->next = *first;
    (*first)->previous = *last;

}

void InsertLast(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = iNo;
    newn->next = NULL;
    newn->previous = NULL;

    if(*first == NULL && *last == NULL)    
    {
        *first = newn;
        *last = newn;
    }

    else                             
    {
        (*last)->next = newn;
        newn->previous = *last; 
        *last = newn;
    }

    (*last)->next = *first;
    (*first)->previous = *last;
    
}

void InsertAtPos(PPNODE first, PPNODE last, int iNo, int iPos)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    int i = 0;
    int iCount = 0;
    iCount = Count(*first,*last);

    //Input filter
    if((iPos < 1) || (iPos > iCount + 1))
    {
        printf("Invalid position");
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(first,last,iNo);
    }

    else if(iPos == iCount + 1)
    {
        InsertLast(first,last,iNo);

    }

    else
    {
        temp = *first;

        newn = (PNODE)malloc(sizeof(NODE));

        newn->data = iNo;
        newn->next = NULL;
        newn->previous = NULL;

        for(i = 1; i < (iPos - 1); i++)
        {
            temp = temp->next;
        }

        newn->next = temp->next;
        temp->next->previous = newn;

        temp->next = newn;
        newn->previous = temp;

    }
    
}

void DeleteFirst(PPNODE first, PPNODE last)
{

    if(*first == NULL && *last == NULL)
    {
        return;
    }

    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }

    else
    {
        *first = (*first)->next;
        free((*first)->previous);

        (*last)->next = *first;
        (*first)->previous = *last;        
    }
    
}

void DeleteLast(PPNODE first, PPNODE last)
{
    
    if(*first == NULL && *last == NULL)
    {
        return;
    }

    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }

    else
    {
        *last = (*last)->previous;
        free((*last)->next);

       (*last)->next = *first;
       (*first)->previous = *last;        
                              
    }    
}

void DeleteAtPos(PPNODE first, PPNODE last, int iPos)
{

    PNODE temp = NULL;

    int i = 0;
    int iCount = 0;
    iCount = Count(*first,*last);

    //Input filter
    if((iPos < 1) || (iPos > iCount))
    {
        printf("Invalid position");
        return;
    }

    if(iPos == 1)
    {
        DeleteFirst(first,last);
    }

    else if(iPos == iCount)
    {
        DeleteLast(first,last);

    }

    else
    {
        temp = *first;

        for(i = 1; i < (iPos - 1); i++)
        {
            temp = temp->next;
        }

        temp->next = temp->next->next;

        free(temp->next->previous);

        temp->next->previous = temp;
    }
        
}

int main()
{
    PNODE head = NULL;  
    PNODE tail = NULL;

    int iRet = 0;            
    int iChoice = 0;
    int iValue = 0;
    int iPosition = 0;

    while(iChoice != 9)
    {
        printf("-----------------------------------------\n");
        printf("Enter your choice : \n");
        printf("-----------------------------------------\n");

        printf("1 : Insert node at first position : \n");
        printf("2 : Insert node at last position : \n");
        printf("3 : Insert node at given position : \n");
        printf("4 : Delete node at first position : \n");
        printf("5 : Delete node at last position : \n");
        printf("6 : Delete node at given position : \n");
        printf("7 : Display elements : \n");
        printf("8 : Number of nodes are in DoublyCL : \n");
        printf("9 : Exit\n");
        printf("-----------------------------------------\n");

        scanf("%d",&iChoice);

        switch(iChoice)
        {
        case 1:
            printf("Enter the value : \n");
            scanf("%d",&iValue);
            InsertFirst(&head,&tail,iValue);
            break;

        case 2:
            printf("Enter the value : \n");
            scanf("%d",&iValue);
            InsertLast(&head,&tail,iValue);
            break;

        case 3:
            printf("Enter the value : \n");
            scanf("%d",&iValue);

            printf("Enter position : \n");
            scanf("%d",&iPosition);

            InsertAtPos(&head,&tail,iValue,iPosition);
            break;

        case 4:
            
            DeleteFirst(&head,&tail);
            break;

        case 5:
            
            DeleteLast(&head,&tail);
            break;

        case 6:

            printf("Enter position : \n");
            scanf("%d",&iPosition);
            
            DeleteAtPos(&head,&tail,iPosition);
            break;

        case 7:

            printf("Elements of Doubly Circular Linked List are : \n");
            Display(head,tail);
            break;

        case 8:

            iRet = Count(head,tail);
            printf("Number of elements are : %d\n",iRet);
            break;

        case 9:

            printf("Thank you for using Doubly Circular Linked List Application\n");
            break;
        
        default:
            printf("Invalid choice\n");
            break;
        }
    }
   
    return 0;
}

File: Data_Structure/Java/DoublyCircularLL.java
// Doubly Circular Linked List
import java.util.Scanner;

class Node
{
    public int data;
    public Node next;
    public Node previous;

    Node(int no)
    {
        this.data = no;
        this.next = null;
        this.previous = null;
    }
}

class DoublyCL 
{
    private Node first;
    private Node last;
    private int iCount;

    public DoublyCL()
    {
        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    public void Display()
    {
        Node temp = first;

        // Input Filter
        if(first == null && last == null)
        {
            return;
        }

        System.out.print(" <=> ");

        do 
        {
            System.out.print("| "+temp.data+" | <=> ");
            temp = temp.next;

        }while(temp != last.next);

        System.out.println();
    }

    public int Count()
    {
        return this.iCount;
    }

    public void InsertFirst(int iNo)
    {
        Node newn = new Node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else 
        {
            newn.next = first;
            first.previous = newn;
            first = newn;
        }

        last.next = first;
        first.previous = last;

        iCount++;
    }

    public void InsertLast(int iNo)
    {
        Node newn = new Node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else 
        {
            last.next = newn;
            newn.previous = last;
            last = newn;
        }

        last.next = first;
        first.previous = last;

        iCount++;
    }

    public void InsertAtPos(int iNo, int iPos)
    {
        Node temp = null;
        Node newn = null;
        int i = 0;

        // Input Filter
        if((iPos < 1) || (iPos > iCount + 1))
        {
            System.out.println("Invalid Position");
            return;
        }

        if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else if(iPos == iCount + 1)
        {
            InsertLast(iNo);
        }
        else 
        {
            temp = first;
            newn = new Node(iNo);

            for(i = 1; i < (iPos - 1); i++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next.previous = newn;

            temp.next = newn;
            newn.previous = temp;

            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else 
        {
            first = first.next;
             
            last.next = first;
            first.previous = last;
        }

        iCount--;
    }

    public void DeleteLast()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else 
        {
            last = last.previous;

            last.next = first;
            first.previous = last;
        }

        iCount--;
        
    }

    public void DeleteAtPos(int iPos)
    {
        Node temp = null;
        int i = 0;

        // Input Filter
        if((iPos < 1) || (iPos > iCount))
        {
            System.out.println("Invalid position");
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == iCount)
        {
            DeleteLast();
        }
        else 
        {
            temp = first;

            for(i = 1; i < (iPos - 1); i++)
            {
                temp = temp.next;
            }

            temp.next = temp.next.next;

            temp.next.previous = temp;

            iCount--;
        }

    }
}

public class DoublyCircularLL 
{
    public static void main(String A[])
    {
        Scanner sobj = new Scanner(System.in);

        DoublyCL dobj = new DoublyCL();

        int iChoice = 0;
        int iValue = 0;
        int iPosition = 0;
        int iRet = 0;

        while(iChoice != 9)
        {
            System.out.println("--------------------------------");
            System.out.println("1 : Insert First");
            System.out.println("2 : Insert Last");
            System.out.println("3 : Insert At Position");
            System.out.println("4 : Delete First");
            System.out.println("5 : Delete Last");
            System.out.println("6 : Delete At Position");
            System.out.println("7 : Display");
            System.out.println("8 : Count");
            System.out.println("9 : Exit");
            System.out.println("--------------------------------");

            System.out.println("Enter your choice : ");
            iChoice = sobj.nextInt();

            switch (iChoice) 
            {
                case 1:
                    System.out.print("Enter Value : ");
                    iValue = sobj.nextInt();
                    dobj.InsertFirst(iValue);
                    break;

                case 2:
                    System.out.print("Enter Value : ");
                    iValue = sobj.nextInt();
                    dobj.InsertLast(iValue);
                    break;

                case 3:
                    System.out.print("Enter Value : ");
                    iValue = sobj.nextInt();
                    System.out.print("Enter Position : ");
                    iPosition = sobj.nextInt();

                    dobj.InsertAtPos(iValue, iPosition);
                    break;

                case 4:
                    dobj.DeleteFirst();
                    break;

                case 5:
                    dobj.DeleteLast();
                    break;

                case 6:
                    System.out.print("Enter position : ");
                    iPosition = sobj.nextInt();

                    dobj.DeleteAtPos(iPosition);
                    break;

                case 7:
                    dobj.Display();
                    break;

                case 8:
                    iRet = dobj.Count();

                    System.out.println("Number of nodes are in Doubly Circular: "+iRet);
                    break;

                case 9:
                    System.out.println("Thank you for using Doubly Circular Linked List Application");
                    break;
                   
                default:
                    System.out.println("Invalid choice");
                    break;
            }
        }
        sobj.close();        
    }
    
}
